File: python/src/sec_edgar_toolkit/core/xbrl/as_reported.py
"""
As-reported statements assembled from the filing's own XBRL fileset.

Combines the instance document (facts, contexts, units) with the
presentation and label linkbases (concept order, hierarchy, display
labels) to produce statements exactly as the company reported them:
properly ordered line items, per-period columns, and dimensional rows
such as segment members.

This works for any filing that carries XBRL, whether or not the SEC
renderer produced ``FilingSummary.xml`` and R-files for it.
"""

from __future__ import annotations

import logging
from typing import Any, Dict, List, Optional

from .instance_document import InstanceDocument, InstanceFact
from .linkbases import LabelLinkbase, PresentationLinkbase, PresentationNode

logger = logging.getLogger(__name__)


class AsReportedStatements:
    """
    Statement access backed by the filing's XBRL fileset.

    Fetches ``<base>_htm.xml`` (or ``<base>.xml`` on older filings),
    ``<base>_pre.xml``, and ``<base>_lab.xml`` from the filing's archive
    folder and assembles ordered line items per statement role.
    """

    def __init__(
        self, archive_base: str, http_client: Any, file_names: List[str]
    ) -> None:
        self._archive_base = archive_base
        self._http = http_client
        self._file_names = file_names

        self._document: Optional[InstanceDocument] = None
        self._presentation: Optional[PresentationLinkbase] = None
        self._labels: Optional[LabelLinkbase] = None
        self._loaded = False
        self._available: Optional[bool] = None

    # ------------------------------------------------------------------
    # Fileset discovery and loading
    # ------------------------------------------------------------------

    def _base_name(self) -> Optional[str]:
        for name in self._file_names:
            if name.lower().endswith(".xsd"):
                return name[:-4]
        return None

    def _instance_name(self) -> Optional[str]:
        base = self._base_name()
        if base is None:
            return None
        for candidate in (f"{base}_htm.xml", f"{base}.xml"):
            if candidate in self._file_names:
                return candidate
        return None

    @property
    def is_available(self) -> bool:
        """True when the filing ships an XBRL fileset we can read."""
        if self._available is None:
            self._available = self._instance_name() is not None
        return self._available

    def _fetch(self, name: str) -> Optional[bytes]:
        try:
            raw = self._http.get_raw(f"{self._archive_base}/{name}")
        except Exception as exc:
            logger.warning(f"Could not fetch {name}: {exc}")
            return None
        if isinstance(raw, str):
            raw = raw.encode("utf-8", errors="ignore")
        return raw

    def _load(self) -> None:
        if self._loaded:
            return
        self._loaded = True

        base = self._base_name()
        instance_name = self._instance_name()
        if base is None or instance_name is None:
            return

        raw_instance = self._fetch(instance_name)
        if raw_instance is None:
            return
        try:
            self._document = InstanceDocument.parse(raw_instance)
        except Exception as exc:
            logger.warning(f"Could not parse instance document: {exc}")
            return

        pre_name = f"{base}_pre.xml"
        if pre_name in self._file_names:
            raw_pre = self._fetch(pre_name)
            if raw_pre is not None:
                try:
                    self._presentation = PresentationLinkbase.parse(raw_pre)
                except Exception as exc:
                    logger.warning(f"Could not parse presentation linkbase: {exc}")

        lab_name = f"{base}_lab.xml"
        if lab_name in self._file_names:
            raw_lab = self._fetch(lab_name)
            if raw_lab is not None:
                try:
                    self._labels = LabelLinkbase.parse(raw_lab)
                except Exception as exc:
                    logger.warning(f"Could not parse label linkbase: {exc}")

    @property
    def document(self) -> Optional[InstanceDocument]:
        """The parsed instance document."""
        self._load()
        return self._document

    @property
    def presentation(self) -> Optional[PresentationLinkbase]:
        """The parsed presentation linkbase."""
        self._load()
        return self._presentation

    @property
    def labels(self) -> Optional[LabelLinkbase]:
        """The parsed label linkbase."""
        self._load()
        return self._labels

    # ------------------------------------------------------------------
    # Statement assembly
    # ------------------------------------------------------------------

    def list_roles(self) -> List[str]:
        """Statement roles defined by the presentation linkbase."""
        presentation = self.presentation
        return list(presentation.roles.keys()) if presentation else []

    def get_statement(self, role: str) -> List[Dict[str, Any]]:
        """
        Assemble one statement in presentation order.

        Line items keep the same shape as the rendered-report parser
        (``concept``, ``label``, ``section``, ``values``, ``units``,
        ``has_values``) plus ``depth``, ``order``, ``abstract``, and
        ``dimensions``. Dimensional facts become their own rows with the
        member labels joined into ``section``.
        """
        document = self.document
        presentation = self.presentation
        if document is None or presentation is None:
            return []

        resolved = presentation.find_role(role)
        if resolved is None:
            return []

        nodes = presentation.ordered_concepts(resolved)
        # Facts are not scoped to roles in the instance, so a concept like
        # revenue carries every dimensional breakdown from every
        # disclosure. The role's own presentation tree names the axes and
        # members it presents, and that set scopes which dimensional rows
        # belong in this statement.
        role_concepts = {node.concept for node in nodes}

        items: List[Dict[str, Any]] = []
        for node in nodes:
            items.extend(self._items_for_node(node, document, role_concepts))
        return items

    def _display_label(self, node: PresentationNode) -> str:
        if self.labels is not None:
            label = self.labels.label_for(node.concept, node.preferred_label)
            if label:
                return label
        return node.concept.split(":")[-1]

    def _member_label(self, member_qname: str) -> str:
        if self.labels is not None:
            label = self.labels.label_for(member_qname)
            if label:
                return label
        return member_qname.split(":")[-1]

    def _items_for_node(
        self,
        node: PresentationNode,
        document: InstanceDocument,
        role_concepts: set,
    ) -> List[Dict[str, Any]]:
        base_label = self._display_label(node)
        facts = document.facts_for(node.concept)

        if not facts:
            # Abstract heading or a concept with no reported facts
            return [
                {
                    "concept": node.concept,
                    "label": base_label,
                    "base_label": base_label,
                    "section": "",
                    "values": {},
                    "units": {},
                    "has_values": False,
                    "abstract": True,
                    "depth": node.depth,
                    "order": node.order,
                    "dimensions": {},
                }
            ]

        # Group the concept's facts by their dimensional signature so the
        # consolidated row and each member breakdown become separate rows.
        grouped: Dict[tuple, List[InstanceFact]] = {}
        for fact in facts:
            context = document.context_of(fact)
            if context is None:
                continue
            signature = tuple(sorted(context.dimensions.items()))
            grouped.setdefault(signature, []).append(fact)

        items: List[Dict[str, Any]] = []
        for signature in sorted(grouped, key=len):
            if any(member not in role_concepts for _dimension, member in signature):
                continue
            group = grouped[signature]
            values: Dict[str, float] = {}
            units: Dict[str, str] = {}
            for fact in group:
                context = document.context_of(fact)
                number = fact.numeric_value()
                if context is None or number is None:
                    continue
                period = context.period_key
                if not period:
                    continue
                values[period] = number
                unit = document.unit_of(fact)
                if unit:
                    units[period] = unit

            if not values:
                continue

            dimensions = dict(signature)
            section = " | ".join(
                self._member_label(member) for member in dimensions.values()
            )
            items.append(
                {
                    "concept": node.concept,
                    "label": f"{section} | {base_label}" if section else base_label,
                    "base_label": base_label,
                    "section": section,
                    "values": values,
                    "units": units,
                    "has_values": True,
                    "abstract": False,
                    "depth": node.depth,
                    "order": node.order,
                    "dimensions": dimensions,
                }
            )
        return items

File: typescript/src/core/xbrl/as-reported.ts
/**
 * As-reported statements assembled from the filing's own XBRL fileset.
 *
 * Combines the instance document (facts, contexts, units) with the
 * presentation and label linkbases (concept order, hierarchy, display
 * labels) to produce statements exactly as the company reported them:
 * properly ordered line items, per-period columns, and dimensional rows
 * such as segment members.
 *
 * This works for any filing that carries XBRL, whether or not the SEC
 * renderer produced `FilingSummary.xml` and R-files for it.
 */

import { StatementLineItem } from "./report-html-parser";
import { InstanceDocument, periodKey } from "./instance-document";
import {
  LabelLinkbase,
  PresentationLinkbase,
  PresentationNode,
} from "./linkbases";

interface RawHttpClient {
  getRaw(url: string): Promise<string>;
}

/** A statement line item with the extra as-reported fields. */
export interface AsReportedLineItem extends StatementLineItem {
  depth: number;
  order: number;
  abstract: boolean;
  dimensions: Record<string, string>;
}

export class AsReportedStatements {
  private readonly archiveBase: string;
  private readonly http: RawHttpClient;
  private readonly fileNames: string[];

  private document: InstanceDocument | null = null;
  private presentation: PresentationLinkbase | null = null;
  private labels: LabelLinkbase | null = null;
  private loaded = false;

  constructor(archiveBase: string, http: RawHttpClient, fileNames: string[]) {
    this.archiveBase = archiveBase;
    this.http = http;
    this.fileNames = fileNames;
  }

  private baseName(): string | null {
    for (const name of this.fileNames) {
      if (name.toLowerCase().endsWith(".xsd")) {
        return name.slice(0, -4);
      }
    }
    return null;
  }

  private instanceName(): string | null {
    const base = this.baseName();
    if (base === null) {
      return null;
    }
    for (const candidate of [`${base}_htm.xml`, `${base}.xml`]) {
      if (this.fileNames.includes(candidate)) {
        return candidate;
      }
    }
    return null;
  }

  get isAvailable(): boolean {
    return this.instanceName() !== null;
  }

  private async fetch(name: string): Promise<string | null> {
    try {
      return await this.http.getRaw(`${this.archiveBase}/${name}`);
    } catch {
      return null;
    }
  }

  private async load(): Promise<void> {
    if (this.loaded) {
      return;
    }
    this.loaded = true;

    const base = this.baseName();
    const instanceName = this.instanceName();
    if (base === null || instanceName === null) {
      return;
    }

    const rawInstance = await this.fetch(instanceName);
    if (rawInstance === null) {
      return;
    }
    try {
      this.document = InstanceDocument.parse(rawInstance);
    } catch {
      return;
    }

    if (this.fileNames.includes(`${base}_pre.xml`)) {
      const rawPre = await this.fetch(`${base}_pre.xml`);
      if (rawPre !== null) {
        try {
          this.presentation = PresentationLinkbase.parse(rawPre);
        } catch {
          this.presentation = null;
        }
      }
    }

    if (this.fileNames.includes(`${base}_lab.xml`)) {
      const rawLab = await this.fetch(`${base}_lab.xml`);
      if (rawLab !== null) {
        try {
          this.labels = LabelLinkbase.parse(rawLab);
        } catch {
          this.labels = null;
        }
      }
    }
  }

  async getDocument(): Promise<InstanceDocument | null> {
    await this.load();
    return this.document;
  }

  async getPresentation(): Promise<PresentationLinkbase | null> {
    await this.load();
    return this.presentation;
  }

  async getLabels(): Promise<LabelLinkbase | null> {
    await this.load();
    return this.labels;
  }

  async listRoles(): Promise<string[]> {
    const presentation = await this.getPresentation();
    return presentation ? Array.from(presentation.roles.keys()) : [];
  }

  /**
   * Assemble one statement in presentation order.
   *
   * Line items keep the same shape as the rendered-report parser plus
   * `depth`, `order`, `abstract`, and `dimensions`. Dimensional facts
   * become their own rows with the member labels joined into `section`.
   */
  async getStatement(role: string): Promise<AsReportedLineItem[]> {
    const document = await this.getDocument();
    const presentation = await this.getPresentation();
    if (document === null || presentation === null) {
      return [];
    }

    const resolved = presentation.findRole(role);
    if (resolved === null) {
      return [];
    }

    const nodes = presentation.orderedConcepts(resolved);
    // Facts are not scoped to roles in the instance; the role's own
    // presentation tree names the axes and members it presents, which
    // scopes the dimensional rows that belong in this statement.
    const roleConcepts = new Set(nodes.map((node) => node.concept));

    const items: AsReportedLineItem[] = [];
    for (const node of nodes) {
      items.push(...this.itemsForNode(node, document, roleConcepts));
    }
    return items;
  }

  private displayLabel(node: PresentationNode): string {
    if (this.labels) {
      const label = this.labels.labelFor(node.concept, node.preferredLabel);
      if (label) {
        return label;
      }
    }
    const colon = node.concept.indexOf(":");
    return colon >= 0 ? node.concept.slice(colon + 1) : node.concept;
  }

  private memberLabel(memberQname: string): string {
    if (this.labels) {
      const label = this.labels.labelFor(memberQname);
      if (label) {
        return label;
      }
    }
    const colon = memberQname.indexOf(":");
    return colon >= 0 ? memberQname.slice(colon + 1) : memberQname;
  }

  private itemsForNode(
    node: PresentationNode,
    document: InstanceDocument,
    roleConcepts: Set<string>,
  ): AsReportedLineItem[] {
    const baseLabel = this.displayLabel(node);
    const facts = document.factsFor(node.concept);

    if (facts.length === 0) {
      return [
        {
          concept: node.concept,
          label: baseLabel,
          baseLabel,
          section: "",
          values: {},
          units: {},
          hasValues: false,
          abstract: true,
          depth: node.depth,
          order: node.order,
          dimensions: {},
        },
      ];
    }

    // Group by dimensional signature: consolidated row plus one row per
    // member combination presented in this role.
    const grouped = new Map<string, typeof facts>();
    for (const fact of facts) {
      const context = document.contextOf(fact);
      if (!context) {
        continue;
      }
      const signature = Object.entries(context.dimensions)
        .sort(([a], [b]) => a.localeCompare(b))
        .map(([dimension, member]) => `${dimension}=${member}`)
        .join("|");
      const list = grouped.get(signature) || [];
      list.push(fact);
      grouped.set(signature, list);
    }

    const items: AsReportedLineItem[] = [];
    const signatures = Array.from(grouped.keys()).sort(
      (a, b) => a.length - b.length,
    );
    for (const signature of signatures) {
      const dimensions: Record<string, string> = {};
      if (signature) {
        let scoped = true;
        for (const pair of signature.split("|")) {
          const [dimension, member] = pair.split("=");
          dimensions[dimension] = member;
          if (!roleConcepts.has(member)) {
            scoped = false;
          }
        }
        if (!scoped) {
          continue;
        }
      }

      const values: Record<string, number> = {};
      const units: Record<string, string> = {};
      for (const fact of grouped.get(signature)!) {
        const context = document.contextOf(fact);
        const number = document.numericValue(fact);
        if (!context || number === null) {
          continue;
        }
        const period = periodKey(context);
        if (!period) {
          continue;
        }
        values[period] = number;
        const unit = document.unitOf(fact);
        if (unit) {
          units[period] = unit;
        }
      }

      if (Object.keys(values).length === 0) {
        continue;
      }

      const section = Object.values(dimensions)
        .map((member) => this.memberLabel(member))
        .join(" | ");
      items.push({
        concept: node.concept,
        label: section ? `${section} | ${baseLabel}` : baseLabel,
        baseLabel,
        section,
        values,
        units,
        hasValues: true,
        abstract: false,
        depth: node.depth,
        order: node.order,
        dimensions,
      });
    }
    return items;
  }
}
